docs/structured-streaming-programming-guide.md

```diff
@@ -671,12 +678,114 @@ windowedCounts = words.groupBy(
 </div>
 
 
+### Handling Late Data and Watermarking
 Now consider what happens if one of the events arrives late to the application.
 For example, a word that was generated at 12:04 but it was received at 12:11. 
-Since this windowing is based on the time in the data, the time 12:04 should be considered for windowing. This occurs naturally in our window-based grouping – the late data is automatically placed in the proper windows and the correct aggregates are updated as illustrated below.
+Since this windowing is based on the time in the data, the time 12:04 should be considered for 
+windowing. This occurs naturally in our window-based grouping – the late data is 
+automatically placed in the proper windows and the correct aggregates are updated as illustrated below.
 
 ![Handling Late Data](img/structured-streaming-late-data.png)
 
+Furthermore, since Spark 2.1, you can define a watermark on the event time, and specify the threshold 
+on how late the date can be in terms of the event time. The engine will automatically track the 
+event time and drop any state that is related to old windows that are not expected to receive older 
+than (max event time seen - late threshold). This allows the engine to bound the size of the state 
+that is needed for calculating windowed aggregates. For example, we can apply watermarking to the 
+previous example as follows.
+
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+
+{% highlight scala %}
+import spark.implicits._
```

Comment:
Just curious, will there be a complete example in the examples folder? In documents like ML, SQL, the code is cited from the example file instead of hard code in the document. Thanks!